### PR TITLE
[serve] Serve get app handle docs

### PR DESCRIPTION
**Pull request by zcin**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zcin** [comment]
Docs build at https://anyscale-ray--39032.com.readthedocs.build/en/39032/serve/deploy-many-models/multi-app.html

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 1]
```diff
@@ -3,65 +3,66 @@
...
+### When to Use Multiple Applications
+Since one application is a unit of upgrade, having multiple applications allows you to deploy multiple models or groups of models that communicate with each other, while still being able to upgrade each application independently.
+
+If you have many independent models each behind different endpoints, and you want to be able to easily add, delete, or upgrade these models, then you should use multiple applications. Each model should then be deployed as a separate application. On the other hand, if you have ML logic and business logic distributed among separate deployments that all need to be executed for a single request, then you should use model composition to build a single application consisting of multiple deployments.
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:15]: The last line here seems a little contradictory to the paragraph above. I think calling one app from another should also be considered model composition, the real comparison is between having deployments in one app vs multiple. Maybe rephrase to say - 

Since one application is a unit of upgrade, having multiple applications allows you to deploy multiple models or groups of models. If you have many independent models each behind different endpoints, and you want to be able to easily add, delete, or upgrade these models, then you should use multiple applications. You can use in-place upgrades <link> to update applications independent of each other.

You can compose deployments across applications. However, if you have one repository with many models and/or business logic that need to be upgraded as a unit, we recommend having all those deployments in one application.

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 2]
```diff
@@ -183,16 +183,19 @@ The command `serve run` blocks the terminal, which allows logs from Serve to str
...
 :::
 
-## New Multi-Application Config
+## Sending requests to applications using Serve handle
+Sometimes, you may want to send a request to an application without using HTTP. For instance, if you want to invoke one application from within another Serve application, it would be inefficient to send the request through HTTP.
+
+For this situation, you can use the Serve API `serve.get_app_handle` to get a handle to any live Serve application. This handle can be used to directly execute a request on an application. For instance
+
+
+## Multi-Application Config
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:198]: Is this repeating stuff above?

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 3]
```diff
@@ -78,66 +79,65 @@ $ serve deploy config.yaml
...
+    last_deployed_time_s: 1693267064.0735464
+    deployments:
+      Translator:
+        status: HEALTHY
+        replica_states:
+          RUNNING: 1
+        message: ''
+```
+
+### Send requests between applications
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:126]: Maybe use FastAPI instead here so it's easy to copy the example

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 4]
```diff
@@ -168,11 +168,11 @@ $ serve run config.yaml
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:168]: Merge serve run section with serve deploy? Should we point out the serve run application change here

File: doc/source/serve/deploy-many-models/multi-app.md (hunk 2 above)
Comment (zcin) [reply doc/source/serve/deploy-many-models/multi-app.md]: Removed!

File: doc/source/serve/deploy-many-models/multi-app.md (hunk 4 above)
Comment (zcin) [reply doc/source/serve/deploy-many-models/multi-app.md:170]: I think it can still be useful to mention it here. I've cut down on the section so it's less repetition, and more of just pointing out you can use serve run with multi app.

File: doc/source/serve/deploy-many-models/multi-app.md (hunk 1 above)
Comment (zcin) [reply doc/source/serve/deploy-many-models/multi-app.md]: I've reworded it a bit, lmk if it looks better!

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 5]
```diff
@@ -3,65 +3,68 @@
...
+### When to Use Multiple Applications
+Many use cases can be solved by either model composition or multi-application. However, both have their own individual benefits and can be used together.
+
+Suppose you have multiple models and/or business logic that all need to be executed for a single request. If they are living in one repository, then most likely they will be upgraded as a unit, so we recommend having all those deployments in one application.
+
+On the other hand, if there are logical groups among these models and/or business logic, e.g. where the groups of models still communicate with each other but live in different repositories, we recommend separating the models into applications. Since one application is a unit of upgrade, having multiple applications allows you to deploy many independent models (or groups of models) each behind different endpoints. You can then use easily add or delete applications from the cluster as well as upgrade applications independently of each other.
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:17]: Maybe add - Another common use-case for multiple applications is separate groups of models that may not communicate with each other but you want to co-host them for improving hardware utilization.

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 6]
```diff
@@ -78,66 +81,65 @@ $ serve deploy config.yaml
...
+```pycon
+>>> requests.post("http://localhost:8000/classify", json={"image_url": "https://cdn.britannica.com/41/156441-050-A4424AEC/Grizzly-bear-Jasper-National-Park-Canada-Alberta.jpg", "should_translate": False}).text
+'Braunbär, Bruin, Ursus arctos'
 ```
 
-:::{note} 
-Notice that in the output of `serve status`, the prefix of each deployment name is the application name. At runtime, all deployments will have their corresponding application prepended to their names.
-:::
 
 ### Inspect Deeper
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:144]: The images here need to be updated

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 7]
```diff
@@ -183,47 +176,17 @@ The command `serve run` blocks the terminal, which allows logs from Serve to str
...
-proxy_location: EveryNode
-http_options:
-  host: 0.0.0.0
-  port: 8000
-```
-
-Then, specify a list of applications to deploy to the Ray cluster. Each application must have a unique name and route prefix.
-```yaml
-applications:
+## Sending requests to applications using Serve handle
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:185]: This seems like a repeat of the section above?

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 8]
```diff
@@ -183,47 +176,17 @@ The command `serve run` blocks the terminal, which allows logs from Serve to str
 
 
 ## Adding, Deleting, and Updating Applications
-The config submitted to the cluster describes the target state for Ray Serve. Consequently, Ray Serve will add, remove, or update based on the list of applications in the config and the config options set for each application.
-* To add an application, add a new entry under the `applications` field.
-* To delete an application, remove the corresponding entry under the `applications` field.
-* To update an application, modify the config options in the corresponding entry under the `applications` field.
+You can add or remove entries under the `applications` field to add or remove applications from the cluster. This will not affect other applications on the cluster. To update an application, modify the config options in the corresponding entry under the `applications` field.
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:179]: add or remove -> add, remove or update

File: doc/source/serve/deploy-many-models/multi-app.md [hunk 9]
```diff
@@ -183,47 +176,17 @@ The command `serve run` blocks the terminal, which allows logs from Serve to str
...
 
 :::{note}
-The update behavior for each application when a config is resubmitted is the same as the old single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
+The update behavior for an application when a config is resubmitted is the same as the old single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
```
Comment (akshay-anyscale) [inline doc/source/serve/deploy-many-models/multi-app.md:182]: Suggest specifying this is for in-place updates so users get a high level idea without having to click on the next page to understand everything

**Comment by akshay-anyscale** [comment]
As discussed yesterday let's remove the Deploy Many Models sub-paging and move Deploy multiple applications and Deploy multiplexed models to the top level

File: doc/source/serve/multi-app.md [hunk 10]
```diff
@@ -157,73 +174,15 @@ As well as the list of replicas for each deployment:
...
-* To add an application, add a new entry under the `applications` field.
-* To delete an application, remove the corresponding entry under the `applications` field.
-* To update an application, modify the config options in the corresponding entry under the `applications` field.
+You can add, remove or update entries under the `applications` field to add, remove or update applications in the cluster. This will not affect other applications on the cluster. To update an application, modify the config options in the corresponding entry under the `applications` field.
 
 :::{note}
-The update behavior for each application when a config is resubmitted is the same as the old single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
+The in-place update behavior for an application when a config is resubmitted is the same as the old single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
```
Comment (akshay-anyscale) [inline doc/source/serve/multi-app.md:182]:
```suggestion
The in-place update behavior for an application when a config is resubmitted is the same as the single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
```

File: doc/source/serve/multi-app.md [hunk 11]
```diff
@@ -78,66 +81,80 @@ $ serve deploy config.yaml
 > Sent deploy request successfully!
```
Comment (akshay-anyscale) [inline doc/source/serve/multi-app.md:81]: Is ray start needed?

File: doc/source/serve/multi-app.md [hunk 12]
```diff
@@ -78,66 +81,80 @@ $ serve deploy config.yaml
...
+$ serve run config.yaml
+> 2023-04-04 11:00:05,901 INFO scripts.py:327 -- Deploying from config file: "config.yaml".
+> 2023-04-04 11:00:07,505 INFO worker.py:1613 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265
+> 2023-04-04 11:00:09,012 SUCC scripts.py:393 -- Submitted deploy config successfully.
 ```
+
+The command `serve run` blocks the terminal, which allows logs from Serve to stream to the console. This helps you test and debug your applications easily. If you want to change your code, you can hit Ctrl-C to interrupt the command and shutdown Serve and all its applications, then rerun `serve run`.
+
+:::{note}
+`serve run` only has support for running multi-application config files. If you want to run applications directly without a config file `serve run` can only run one application at a time.
```
Comment (akshay-anyscale) [inline doc/source/serve/multi-app.md:105]: I think this note is a little unclear. serve run does allow you to specify applications right? Maybe explain that when using serve run it'll restart everything

**Review by akshay-anyscale — APPROVED** [review]
approved with some small comments

File: doc/source/serve/multi-app.md (hunk 11 above)
Comment (zcin) [reply doc/source/serve/multi-app.md:81]: It's needed before running serve deploy. Do you mean if it's necessary to include in the docs?

File: doc/source/serve/multi-app.md (hunk 12 above)
Comment (zcin) [reply doc/source/serve/multi-app.md:105]: `serve run` can't run multiple applications by import path. I've updated the note to make that more clear!

File: doc/source/serve/multi-app.md [hunk 13]
```diff
@@ -106,50 +116,53 @@ Check the status of the applications by running `serve status`.
 ```console
 $ serve status
 proxies:
-  59e6321aa1699fa09ed1094e1f0b3c37afd47fc208802db3c07328d0: HEALTHY
+  2e02a03ad64b3f3810b0dd6c3265c8a00ac36c13b2b0937cbf1ef153: HEALTHY
 applications:
-  app1:
+  classify:
```
Comment (GeneDer) [inline doc/source/serve/multi-app.md:121]: I think the config above is still using "app1" and "app2"?

File: doc/source/serve/multi-app.md [hunk 14]
```diff
@@ -49,24 +55,20 @@ grpc_options:
   grpc_servicer_functions: []
 
 applications:
-
 - name: app1
```
Comment (GeneDer) [inline doc/source/serve/multi-app.md:58]: Should we also change the app name to "classify" and "translate"?

File: doc/source/serve/multi-app.md [hunk 15]
```diff
@@ -106,50 +116,53 @@ Check the status of the applications by running `serve status`.
...
+:start-after: __serve_example_modified_begin__
+:end-before: __serve_example_modified_end__
+```
+
+Then, sending requests to the classifier application with the `should_translate` flag set to True:
+```python
+import requests
+
+requests.post("http://localhost:8000/classify", json={"image_url": "https://cdn.britannica.com/41/156441-050-A4424AEC/Grizzly-bear-Jasper-National-Park-Canada-Alberta.jpg", "should_translate": False}).text
+'Braunbär, Bruin, Ursus arctos'
```
Comment (GeneDer) [inline doc/source/serve/multi-app.md:162]: I'm kinda confused, should this be in English since `should_translate` is false?

File: doc/source/serve/multi-app.md [hunk 16]
```diff
@@ -83,21 +85,29 @@ $ serve deploy config.yaml
 > Sent deploy request successfully!
 ```
 
-Query the applications at their respective endpoints, `/calculator` and `/greet`.
-```pycon
->>> requests.post("http://localhost:8000/calculator", json=["ADD", 5]).json()
-7
+Query the applications at their respective endpoints, `/classify` and `/translate`.
+```python
```
Comment (GeneDer) [inline doc/source/serve/multi-app.md:89]: Should use [testcode](https://docs.ray.io/en/master/ray-contribute/writing-code-snippets.html#code-output-style-examples) here. But maybe better to just move those to the .py file. 

Same goes to the rest of this file.

File: doc/source/serve/multi-app.md (hunk 15 above)
Comment (GeneDer) [reply doc/source/serve/multi-app.md:162]: Or maybe `should_translate` is meant to be true?

File: doc/source/serve/doc_code/image_classifier_example.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+from translator_example import app as translator_app  # noqa
+
+serve.run(
+    ModifiedImageClassifier.bind(downloader.bind()),
+    name="app1",
+    route_prefix="/classify",
+)
+serve.run(translator_app, name="app2")
```
Comment (GeneDer) [inline doc/source/serve/doc_code/image_classifier_example.py:84]: Also just want to note in the code those apps are named "app1" and "app2"

File: doc/source/serve/multi-app.md (hunk 15 above)
Comment (zcin) [reply doc/source/serve/multi-app.md:162]: Yeah it should be true, fixed

File: doc/source/serve/doc_code/image_classifier_example.py (hunk 17 above)
Comment (zcin) [reply doc/source/serve/doc_code/image_classifier_example.py:84]: Yeah they should be app1 and app2, the serve status output was wrong

File: doc/source/serve/multi-app.md (hunk 16 above)
Comment (zcin) [reply doc/source/serve/multi-app.md]: Changed to using test code!

File: doc/source/serve/multi-app.md (hunk 13 above)
Comment (zcin) [reply doc/source/serve/multi-app.md:123]: Yeah this output is wrong, I've changed it to app1 and app2

File: doc/source/serve/multi-app.md (hunk 14 above)
Comment (zcin) [reply doc/source/serve/multi-app.md:58]: Since we include `serve build` as part of this doc and it autogenerates `app1` and `app2`, I think it's OK to leave as `app1` and `app2`

File: doc/source/_toc.yml [hunk 18]
```diff
@@ -188,10 +188,8 @@ parts:
           - file: serve/key-concepts
           - file: serve/develop-and-deploy
           - file: serve/model_composition
-          - file: serve/deploy-many-models/index
-            sections:
```
Comment (edoakes) [inline doc/source/_toc.yml:192]: nice, I was going to suggest this :)

**Review by edoakes — APPROVED** [review]

File: doc/source/serve/multi-app.md [hunk 19]
```diff
@@ -83,21 +85,31 @@ $ serve deploy config.yaml
...
 
-:::{tip}
-If you prefer to use `cURL` to ping these endpoints, add the `-L` flag. The `DAGDriver` does an HTTP redirect, and the flag ensures that `cURL` follows the redirect:
+The command `serve run` blocks the terminal, which allows logs from Serve to stream to the console. This helps you test and debug your applications easily. If you want to change your code, you can hit Ctrl-C to interrupt the command and shutdown Serve and all its applications, then rerun `serve run`.
 
-```
-curl -L -X POST -H "Content-Type: application/json" -d '["ADD", 5]' localhost:8000/calculator
-```
+:::{note}
+`serve run` only has support for running multi-application config files. If you want to run applications by directly passing in an import path, `serve run` can only run one application import path at a time.
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:112]:
```suggestion
`serve run` only supports running multi-application config files. If you want to run applications by directly passing in an import path, `serve run` can only run one application import path at a time.
```

File: doc/source/serve/multi-app.md [hunk 20]
```diff
@@ -106,50 +118,52 @@ Check the status of the applications by running `serve status`.
...
+### Send requests between applications
+You can also make calls between applications without going through HTTP by using the Serve API `serve.get_app_handle` to get a handle to any live Serve application on the cluster. This handle can be used to directly execute a request on an application. Take the classifier and translator app above as an example. We can modify the `__call__` method of the `ImageClassifier` to check for another parameter in the HTTP request, and send requests to the translator application.
+
+```{literalinclude} doc_code/image_classifier_example.py
+:language: python
+:start-after: __serve_example_modified_begin__
+:end-before: __serve_example_modified_end__
+```
+
+Then, sending requests to the classifier application with the `should_translate` flag set to True:
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:159]:
```suggestion
Then, send requests to the classifier application with the `should_translate` flag set to True:
```

File: doc/source/serve/multi-app.md [hunk 21]
```diff
@@ -106,50 +118,52 @@ Check the status of the applications by running `serve status`.
...
+```{literalinclude} doc_code/image_classifier_example.py
+:language: python
+:start-after: __second_request_begin__
+:end-before: __second_request_end__
+```
+
+
 ### Inspect Deeper
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:167]:
```suggestion
### Inspect deeper
```

File: doc/source/serve/multi-app.md [hunk 22]
```diff
@@ -168,73 +182,15 @@ As well as the list of replicas for each deployment:
...
-```
-
-The command `serve run` blocks the terminal, which allows logs from Serve to stream to the console. This helps you test and debug your applications easily. If you want to change your code, you can hit Ctrl-C to interrupt the command and shutdown Serve and all its applications, then rerun `serve run`.
-
-:::{note}
-`serve run` only has support for running multi-application config files. If you want to run applications directly without a config file `serve run` can only run one application at a time.
-:::
-
 
 ## Adding, Deleting, and Updating Applications
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:186]:
```suggestion
## Add, delete, and update applications
```

File: doc/source/serve/multi-app.md [hunk 23]
```diff
@@ -168,73 +182,15 @@ As well as the list of replicas for each deployment:
...
-The config submitted to the cluster describes the target state for Ray Serve. Consequently, Ray Serve will add, remove, or update based on the list of applications in the config and the config options set for each application.
-* To add an application, add a new entry under the `applications` field.
-* To delete an application, remove the corresponding entry under the `applications` field.
-* To update an application, modify the config options in the corresponding entry under the `applications` field.
+You can add, remove or update entries under the `applications` field to add, remove or update applications in the cluster. This will not affect other applications on the cluster. To update an application, modify the config options in the corresponding entry under the `applications` field.
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:187]:
```suggestion
You can add, remove, or update entries under the `applications` field to add, remove, or update applications in the cluster. This doesn't affect other applications on the cluster. To update an application, modify the config options in the corresponding entry under the `applications` field.
```

File: doc/source/serve/multi-app.md [hunk 24]
```diff
@@ -168,73 +182,15 @@ As well as the list of replicas for each deployment:
...
 
 :::{note}
-The update behavior for each application when a config is resubmitted is the same as the old single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
+The in-place update behavior for an application when a config is resubmitted is the same as the single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:190]:
```suggestion
The in-place update behavior for an application when you resubmit a config is the same as the single-application behavior. For how an application reacts to different config changes, see [Updating a Serve Application](serve-inplace-updates).
```

File: doc/source/serve/multi-app.md [hunk 25]
```diff
@@ -168,73 +182,15 @@ As well as the list of replicas for each deployment:
...
-  route_prefix: /greet
-  import_path: greet:app
-  runtime_env: {}
-  deployments:
-  - name: greet
-  - name: DAGDriver
-```
 
 (serve-config-migration)=
 ### Migrating from a Single-Application Config
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:196]:
```suggestion
### Migrating from a single-application config
```

File: doc/source/serve/multi-app.md [hunk 26]
```diff
@@ -3,37 +3,43 @@
 
 In Ray 2.4+, deploying multiple independent Serve applications is supported. This user guide walks through how to generate a multi-application config file and deploy it using the Serve CLI, and monitor your applications using the CLI and the Ray Serve dashboard.
```
Comment (angelinalg) [inline doc/source/serve/multi-app.md:4]:
```suggestion
Ray 2.4+ supports deploying multiple independent Serve applications. This user guide walks through how to generate a multi-application config file and deploy it using the Serve CLI, and monitor your applications using the CLI and the Ray Serve dashboard.
```

**Review by angelinalg — APPROVED** [review]
Suggest that you use sentence case for all sub-headings/titles. I tried to catch them all but may have missed some.

**Review by GeneDer — APPROVED** [review]
LGTM!

File: doc/source/serve/multi-app.md [hunk 27]
```diff
@@ -1,23 +1,23 @@
 (serve-multi-application)=
 # Deploy Multiple Applications
 
-In Ray 2.4+, deploying multiple independent Serve applications is supported. This user guide walks through how to generate a multi-application config file and deploy it using the Serve CLI, and monitor your applications using the CLI and the Ray Serve dashboard.
+Ray 2.4+ supports deploying multiple independent Serve applications. This user guide walks through how to generate a multi-application config file and deploy it using the Serve CLI, and monitor your applications using the CLI and the Ray Serve dashboard.
```
Comment (edoakes) [inline doc/source/serve/multi-app.md:4]: we can drop the release specification now

**Comment by zcin** [comment]
@edoakes This change is ready to merge!
